photon-lib/src/main/java/org/photonvision/simulation/VideoSimUtil.java

```
@@ -0,0 +1,335 @@
+        }
+
+        kTag16h5MarkerPts = get16h5MarkerPts();
+    }
+
+    /** Updates the properties of this CvSource video stream with the given camera properties. */
+    public static void updateVideoProp(CvSource video, SimCameraProperties prop) {
+        video.setResolution(prop.getResWidth(), prop.getResHeight());
+        video.setFPS((int) prop.getFPS());
+    }
+
+    /**
+     * Gets the points representing the corners of this image. Because image pixels are accessed
+     * through a Mat, the point (0,0) actually represents the center of the top-left pixel and not the
+     * actual top-left corner.
+     *
+     * @param size Size of image
+     */
+    public static final Point[] getImageCorners(Size size) {
+        return new Point[] {
+            new Point(-0.5, -0.5),
+            new Point(size.width - 0.5, -0.5),
+            new Point(size.width - 0.5, size.height - 0.5),
+            new Point(-0.5, size.height - 0.5)
+        };
+    }
+
+    /**
+     * Gets the 8x8 (grayscale) image of a specific 16h5 AprilTag.
+     *
+     * @param id The fiducial id of the desired tag
+     */
+    public static Mat get16h5TagImage(int id) {
+        String name = kTag16h5ImageName;
+        String idString = String.valueOf(id);
+        name = name.substring(0, name.length() - idString.length()) + idString;
+
+        var resource = VideoSimUtil.class.getResource(kResourceTagImagesPath + name + ".png");
+        // local IDE tests
+        String path = kLocalTagImagesPath + name + ".png";
+        // gradle tests
+        if (resource != null) {
+            path = resource.getPath();
+            if (path.startsWith("/")) path = path.substring(1);
+        }
+        Mat result = new Mat();
+        if (!path.startsWith("file")) result = Imgcodecs.imread(path, Imgcodecs.IMREAD_GRAYSCALE);
+        // reading jar file
+        if (result.empty()) {
+            BufferedImage buf;
+            try {
+                buf = ImageIO.read(resource);
```

no null check here? Is this assumed to be non-null based on the previous check?

---

if you mean `result`, it should always be at least an empty Mat. If you mean `buf`, I believe it should throw an exception when it can't read the desired image.

---

I mean resource, you bill check before but not beyond that if statement

---

If `resource` is null, `ImageIO.read` will throw an exception.